features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/MetaTagDataLoader.java

```diff
@@ -0,0 +1,220 @@
+    private MetaTagConfiguration config;
+
+    @Inject
+    public MetaTagDataLoader(final NodeDao nodeDao, final SessionUtils sessionUtils) {
+        this.nodeDao = nodeDao;
+        this.sessionUtils = sessionUtils;
+        setConfig(new MetaTagConfiguration(Maps.fromProperties(System.getProperties())));
+    }
+
+    public void setConfig(final MetaTagConfiguration config) {
+        this.config = config;
+    }
+
+    public Map<String, String> load(final CollectionResource resource) {
+        return sessionUtils.withReadOnlyTransaction(() -> {
+            final Map<String, String> tags = new HashMap<>();
+            addTag(tags, MetaTagKey.resourceLabel, () -> Optional.ofNullable(resource.getInterfaceLabel()).orElse(null));
+
+            if (resource.getResourceTypeName().equals(CollectionResource.RESOURCE_TYPE_NODE)) {
+                String nodeCriteria = getNodeCriteriaFromResource(resource);
+                mapNode(tags, nodeCriteria);
+            } else if (resource.getResourceTypeName().equals(CollectionResource.RESOURCE_TYPE_IF)) {
+
+                String nodeCriteria = getNodeCriteriaFromResource(resource);
+                if (!Strings.isNullOrEmpty(nodeCriteria)) {
+                    mapNode(tags, nodeCriteria);
+                }
+            } else if (resource.getResourceTypeName().equals(CollectionResource.RESOURCE_TYPE_LATENCY)) {
+                mapResponseTimeResource(resource, tags);
+            } else {
+                String nodeCriteria = getNodeCriteriaFromResource(resource);
+                if (!Strings.isNullOrEmpty(nodeCriteria)) {
+                    mapNode(tags, nodeCriteria);
+                }
+            }
+
+            return tags;
+        });
+    }
+
+    private void mapNode(final Map<String, String> tags, final String nodeCriteria) {
+        Optional<OnmsNode> node = getNode(nodeCriteria);
+        addTag(tags, MetaTagKey.nodeCriteria, () -> nodeCriteria);
+        addTag(tags, MetaTagKey.nodeLabel, () -> node.map(OnmsNode::getLabel).orElse(null));
+        addTag(tags, MetaTagKey.location, () -> node.map(OnmsNode::getLocation).map(OnmsMonitoringLocation::getLocationName).orElse(null));
+        addTag(tags, MetaTagKey.sysObjectID, () -> node.map(OnmsNode::getSysObjectId).orElse(null));
+        addTag(tags, MetaTagKey.foreignSource, () -> node.map(OnmsNode::getForeignSource).orElse(null));
+        addTag(tags, MetaTagKey.foreignId, () -> node.map(OnmsNode::getForeignId).orElse(null));
+        // TODO Patrick: map rest of attributes
+
+        // categories
+        if(node.isPresent()) {
+            AtomicInteger n = new AtomicInteger(1);
+            // Tags are only unique by key and value
+            node.get().getCategories().stream()
+                    .map(OnmsCategory::getName)
+                    .filter(config::isCategoryEnabled)
+                    // TODO: Patrick: is this acceptable? Or would we rather create Tags instead of a HashMap?
```

Comment:
Tags are unique by the combination of key and value. Thus we could have category tags like:
Tag(key="category", value="myFirstCategory"), Tag(key="category", value="mySecondCategory") etc.
The problem here is that we store the future tags as key value pairs in a Map<String, String>. That disallows us to store the tags with the same key since they have to be unique. I see two ways out of that:

1. make the keys unique like: Tag(key="category_1", value="myFirstCategory"), Tag(key="category_2", value="mySecondCategory")
2. Enhance (by our own class) or substitute org.opennms.newts.api.Resource to store a set of Tags instead a Map of Strings

The first is the easier way (currently implemented) the second one more work but I guess the better way. I am not sure how we use the category tags later so it's hard for me to make a call here. @j-white What is your take?

Reply:
I was thinking we could do something like this:
```
for(String categoryName : categoryNames) {
  if (!enabledCategoryNames.contains(categoryName)) {
      continue;
   }
  addTag(tags, "cat" + categoryName, categoryName);
}
```

Reply:
:+1: fixed